### PR TITLE
[FIX] export AMS result as valid mzTab

### PR DESCRIPTION
**Pull request by timosachsenberg**

### CODE REVIEW

**Comment by timosachsenberg** [comment]
should address https://github.com/OpenMS/OpenMS/issues/1197

**Comment by timosachsenberg** [comment]
will amend as I found another bug

**Comment by timosachsenberg** [comment]
renamed double getAdductMass()  to getObservedMZ()
renamed double getErrorPPM to getDBErrorPPM() as this is the error for the uncharged, 1-mer adductless small molecule 
getErrorPPM is the error between observed mz and theoretical mz (nmer, with adduct and charge)
- Errors are calculated relative to theoretical masses

**Comment by timosachsenberg** [comment]
renamed queryByMass to queryByMZ as an observed m/z value is passed

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp [hunk 1]
```diff
@@ -450,11 +489,11 @@ namespace OpenMS
 
   void AccurateMassSearchResult::outputResults() const
   {
-    std::cout << "adduct_mass: " << std::setprecision(8) << adduct_mass_ << "\n";
+    std::cout << "adduct_mass: " << std::setprecision(8) << observed_mz_ << "\n";
```
Comment (FabianAicheler) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: Rename to "observed_mz: "

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp [hunk 2]
```diff
@@ -797,7 +844,7 @@ namespace OpenMS
       hit.setMetaValue("charge", it_row->getCharge());
       hit.setMetaValue("modifications", it_row->getFoundAdduct());
       hit.setMetaValue("chemical_formula", it_row->getFormulaString());
-      hit.setMetaValue("ppm_error", it_row->getErrorPPM());
+      hit.setMetaValue("ppm_error", it_row->getDBErrorPPM());
```
Comment (FabianAicheler) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: Rename to 'ppm_db_error' ?

Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: good point... If @cbielow is fine with the rename I will include it

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp (hunk 1 above)
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: agree. will rename to "observed MZ:"

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp (hunk 2 above)
Comment (FabianAicheler) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: Didn't  notice anything besides that, from what I can see 
+1

File: src/openms/include/OpenMS/ANALYSIS/ID/AccurateMassSearchEngine.h [hunk 3]
```diff
@@ -68,29 +68,64 @@ namespace OpenMS
...
+
+    /// set the error between found and query mass in ppm
+    void setDBErrorPPM(const double);
+
+    /// get the error between observed and theoretical m/z in ppm
     double getErrorPPM() const;
-    void setErrorPPM(const double&);
 
+    /// set the error between observed and theoretical m/z in ppm
+    void setErrorPPM(const double);
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/ID/AccurateMassSearchEngine.h]: maybe setMZErrorPPM() ?

File: src/openms/include/OpenMS/ANALYSIS/ID/AccurateMassSearchEngine.h [hunk 4]
```diff
@@ -246,8 +283,12 @@ namespace OpenMS
 
       AdductInfo_(const String& name, const EmpiricalFormula& adduct, int charge, bool is_intrinsic = false, uint mol_multiplier = 1);
 
+      /// returns the neutral mass of the small molecule without adduct (creates monomer fron nmer, decharges and removes the adduct (given m/z of [nM+Adduct]/|charge| returns mass of [M])
       double getNeutralMass(double observed_mz) const;
 
+      /// returns the m/z of the small molecule with neutral mass @p neutral_mass if the adduct is added (given mass of [M] returns m/z of [nM+Adduct]/|charge|)
+      double getMZ(double neutral_mass) const;
+
       /// checks if an adduct (e.g.a 'M+2K-H;1+') is valid, i.e.a if the losses (==negative amounts) can actually be lost by the compound given in @p db_entry.
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/ID/AccurateMassSearchEngine.h]: typo -- could you fix it please?

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp [hunk 5]
```diff
@@ -84,6 +84,19 @@ namespace OpenMS
...
+    if (!is_intrinsic_)
+    { // correct for electron masses
+      double electrons_mass_diff(charge_ * Constants::ELECTRON_MASS_U);
+      neutral_nmer_mass_with_adduct += electrons_mass_diff;
+    }
+
+    return neutral_nmer_mass_with_adduct / abs(charge_);
+  }
+
   /// checks if an adduct (e.g.a 'M+2K-H;1+') is valid, i.e.a if the losses (==negative amounts) can actually be lost by the compound given in @p db_entry.
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: typo...

File: src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp [hunk 6]
```diff
@@ -797,7 +844,7 @@ namespace OpenMS
       hit.setMetaValue("charge", it_row->getCharge());
       hit.setMetaValue("modifications", it_row->getFoundAdduct());
       hit.setMetaValue("chemical_formula", it_row->getFormulaString());
-      hit.setMetaValue("ppm_error", it_row->getErrorPPM());
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: why not store mz-error as well?

Comment (cbielow) [inline src/openms/source/ANALYSIS/ID/AccurateMassSearchEngine.cpp]: why not store m/z error as well?

**Comment by timosachsenberg** [comment]
I adressed the minor additions. @cbielow  can you check if after the merge the annotation in TOPPView works correctly?

**Comment by hroest** [comment]
Did wrap.the new.functions in pyopenms?

sent from my smartphone -- you may keep any typos you find.

**Comment by timosachsenberg** [comment]
Yep. I did not run the pyOpenMS build though

**Comment by timosachsenberg** [comment]
@cbielow I was a bit surprised that db ppm error and m/z ppm error differ sometimes quite alot (see the updated featureXML file) in some cases. In other cases they are basically identical. I did not find an obvious bug by looking at it. Do you have an explanation?

**Comment by hroest** [comment]
After pyOpenMS failed, could you try fixing the wrappers
http://cdash.openms.de/viewBuildError.php?buildid=119543 ?

sent from my smartphone -- you may keep any typos you find.

**Comment by timosachsenberg** [comment]
should be fixed in https://github.com/OpenMS/OpenMS/pull/1228

**Pull request merged**